service-mvt/src/main/java/org/oskari/service/mvt/SimpleFeaturesMVTEncoder.java

```diff
@@ -0,0 +1,298 @@
+                if (g == null) {
+                    continue;
+                }
+
+                Geometry geom = (Geometry) g;
+                geom.setUserData(sf);
+                geoms.add(geom);
+            }
+        }
+
+        return JtsAdapter.createTileGeom(geoms, tileEnvelope, clipEnvelope, GF,
+                new MvtLayerParams(512, extent), new GeomMinSizeFilter(6, 6)).mvtGeoms;
+    }
+
+    public static List<Geometry> asMVTGeoms(SimpleFeatureCollection sfc, double[] bbox, int extent, int buffer) {
+        Envelope tileEnvelope = new Envelope(bbox[0], bbox[2], bbox[1], bbox[3]);
+        Envelope clipEnvelope = new Envelope(tileEnvelope);
+        if (buffer > 0) {
+            double bufferSizePercent = (double) buffer / extent;
+            double dx = bufferSizePercent * tileEnvelope.getWidth();
+            double dy = bufferSizePercent * tileEnvelope.getHeight();
+            clipEnvelope.expandBy(dx, dy);
+        }
+        Geometry tileEnvelopeGeom = GF.toGeometry(tileEnvelope);
+        Geometry tileClipGeom = GF.toGeometry(clipEnvelope);
+        RectangleIntersects rectIntersects = new RectangleIntersects((Polygon) tileEnvelopeGeom);
+
+        double res = tileEnvelope.getWidth() / extent;
+        double resHalf = res / 2;
+
+        double tx = tileEnvelope.getMinX();
+        double ty = tileEnvelope.getMaxY();
+        double sx = (double) extent / tileEnvelope.getWidth();
+        double sy = -((double) extent / tileEnvelope.getHeight());
+
+        ToMVTSpace snapToGrid = new ToMVTSpace(tx, ty, sx, sy);
+        
+        GeometryEditor editor = new GeometryEditor(GF);
+
+        List<Geometry> mvtGeoms = new ArrayList<>();
+        try (SimpleFeatureIterator it = sfc.features()) {
+            while (it.hasNext()) {
+                SimpleFeature sf = it.next();
+                Object g = sf.getDefaultGeometry();
+                if (g == null) {
+                    continue;
+                }
+                Geometry geom = (Geometry) g;
+                if (geom.isEmpty()) {
+                    continue;
+                }
+                Envelope geomEnvelope = geom.getEnvelopeInternal();
+
+                // Quickly check that atleast the envelopes intersect
+                if (!tileEnvelope.intersects(geom.getEnvelopeInternal())) {
+                    continue;
+                }
+
+                if (geom instanceof LineString
+                        || geom instanceof Polygon
+                        || geom instanceof MultiLineString
+                        || geom instanceof MultiPolygon) {
+                    double bboxWidth = geomEnvelope.getWidth();
+                    double bboxHeight = geomEnvelope.getHeight();
+                    if (bboxWidth * bboxWidth + bboxHeight * bboxHeight < resHalf * resHalf) {
+                        continue;
+                    }
+                }
+
+                // geom = editor.edit(geom, removeRepeatedPoints);
```

Comment:
To make the loop more compact, everything up to this point could possible be moved to some method like:

     Geometry geom = validate(sf.getDefaultGeometry(), tileEnvelope)